text/0000-generalized-index.md

```diff
@@ -0,0 +1,236 @@
+
+```rust
+fn main() {
+  let range = X[..];
+}
+```
+
+This change is a *small change* in regard to newcomers – it’s more general than the legacy trait but
+it doesn’t change the way people use it. However, for people used to this trait and especially for
+developers who’ve already implemented this trait for custom types, code breakage must be addressed
+as the following: because the legacy trait’s implementors return only references, we can use
+`rustfix` to automatically migrate to the new trait. For people wanting to do it manually, the
+following guide should be sufficient:
+
+  1. Change the type of the `impl` signature to add the `'a` lifetime.
+  2. Change the `Index::Output` type by prepending a `&'a` in front of the type defined in there.
+  3. Change the definition of the `Index::index` function.
+  4. That’s all.
+
+This change then only changes the type signatures. The impact on the implementations’ bodies is
+null, because the new trait is more general than the legacy one (it creates a superset of accepted
+types, so the legacy types are in this bigger set).
+
+# Reference-level explanation
+[reference-level-explanation]: #reference-level-explanation
+
+Most of the current implementors must receive the fix described in the end of section
+[guide-level-explanation]. This will change the documentation (but not the actual implementations).
+
+# Drawbacks
+[drawbacks]: #drawbacks
+
+## Breakage
+
+This change has the huge drawback that it will generate breakage in most codebases that:
+
+ - `impl Index<_> for` any custom types.
+ - The std / core codebase will break as well since most common implementors are defined in there.
+
+## More general yet more complex trait
+
+The new trait version is more general yet more complex to wrap your fingers around. The `'a`
+lifetime exposed in the trait interface might confuse people and could be seen has a drawback for
+most of the common cases – i.e. when people just want to index the **inside of an object**.
+
+# Rationale and alternatives
+[alternatives]: #alternatives
+
+This change is small in terms of code modification to operate on codebase and doesn’t introduce any
+newkeyword or language construct. Another lead to implement this feature would be to add a new
+operator, like the `[[]]` operator or even `[move idx]` construct. This would then require adding a
+new trait, like `IndexMove`, to make the whole design work.
+
+## How not to introduce a breaking change
+
+In order to mitigate the breaking change, we could add a new trait, such as `NewIndex`, make the
+`[]` operator uses this trait and provide the following `default` universal implementation:
+
+```rust
+impl<'a, T, Idx> NewIndex<Idx> for T where T: Index<Idx> {
+  type NewOutput = &'a T::Output;
+
+  default fn new_index(&'a self, idx: Idx) -> Self::NewOutput {
+    self.index(idx)
+  }
+}
+```
+
+# Unresolved questions
+[unresolved]: #unresolved-questions
+
+  - Should this RFC also concern `IndexMut`?
```

Comment:
Yes, this RFC also has to cover `IndexMut` - that trait extends `Index` and reuses its output type. That setup will not work with this proposed change.